include/mbedtls/psa_util.h

```diff
@@ -419,4 +419,88 @@ static inline int mbedtls_psa_tls_ecpoint_to_psa_ec( unsigned char const *src,
 
 #endif /* MBEDTLS_USE_PSA_CRYPTO */
 
+/* Expose whatever RNG the PSA subsystem uses to applications using the
+ * mbedtls_xxx API. The declarations here need to be consistent with the
+ * implementation in library/psa_crypto_random_impl.h. */
+#if defined(MBEDTLS_PSA_CRYPTO_C)
+
+/* The type of a `f_rng` random generator function that many library functions
+ * take.
+ *
+ * This type name is not part of the Mbed TLS stable API. It may be renamed
+ * or moved without warning.
+ */
+typedef int mbedtls_f_rng_t( void *p_rng, unsigned char *output, size_t output_size );
+
+#if defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+
+/** The random generator function for the PSA subsystem.
+ *
+ * This function is suitable as the `f_rng` random generator function
+ * parameter of many `mbedtls_xxx` functions. It must be used in combination
+ * with the random generator state #MBEDTLS_PSA_RANDOM_STATE.
+ *
+ * The implementation of this function depends on the configuration of the
+ * library.
+
+ * \note Depending on the configuration, this may be a function or
+ *       a pointer to a function.
+ *
+ * \note This function may only be used if the PSA crypto subsystem is active.
+ *       This means that you must call psa_crypto_init() before any call to
+ *       this function, and you must not call this function after calling
+ *       mbedtls_psa_crypto_free().
+ *
+ * \param p_rng         The random generator context. This must be
+ *                      #MBEDTLS_PSA_RANDOM_STATE. No other state is
+ *                      supported.
+ * \param output        The buffer to fill. It must have room for
+ *                      \c output_size bytes.
+ * \param output_size   The number of bytes to write to \p output.
+ *                      This function may fail if \p output_size is too
+ *                      large. It is guaranteed to accept any output size
+ *                      requested by Mbed TLS library functions. The
+ *                      maximum request size depends on the library
+ *                      configuration.
+ *
+ * \return              \c 0 on success.
+ * \return              An `MBEDTLS_ERR_ENTROPY_xxx`,
+ *                      `MBEDTLS_ERR_CTR_DRBG_xxx` or
+ *                      `MBEDTLS_ERR_HMAC_DRBG_xxx` on error.
+ */
+int mbedtls_psa_get_random( void *p_rng,
```

Comment:
Isn't this a straight-up copy of what's in `psa_crypto_random_impl.h`? If this is meant to replace that, why isn't the other file being removed?

Reply:
These are just the declarations that need to be in public headers. `psa_crypto_random_impl.h` has more declarations and definitions.

There's some duplication between the two headers, but I think it's the minimum necessary, or at least the minimum advisable. A couple of things are duplicated, in a way that a compiler will complain if they diverge, and this way all the implementation stuff is in `psa_crypto_random_impl.h` where it's easy to see all in one place, but we still get the compiler to complain if the implementation no longer matches the public interface.

Reply:
Are you convinced by my explanation, or do some things need to be removed from `psa_crypto_random_impl.h`?

Reply:
Once again I'm not sure what change you expect. To me, the comments at the top of the files referring to each other are sufficient. Do you want this to be also mentioned in the documentation of each macro/function/type?

Reply:
From your comment above the reason of the duplication seems to be:
"A couple of things are duplicated, in a way that a compiler will complain if they diverge, and this way all the implementation stuff is in psa_crypto_random_impl.h where it's easy to see all in one place, but we still get the compiler to complain if the implementation no longer matches the public interface."
I'm just saying what about to put something equivalent as documentation in the header files.

Reply:
Revisiting this, the compiler doesn't enforce consistency: there's duplication here that's rejected if both headers are included (`static` variable definition, type definition). So I do need to remove some things from `psa_crypto_random_impl.h`.

Reply:
I've removed the redundant definitions and kept redundant declarations where possible. I kept the duplication of the definition of the macro `MBEDTLS_PSA_RANDOM_STATE` because this is both permitted and checked.